CHANGELOG.md

```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `Surrogate` base class now exposes a `to_botorch` method
 - `SubspaceDiscrete.to_searchspace` and `SubspaceContinuous.to_searchspace`
   convenience constructor
+- `register_hooks` utility enabling user-defined augmentation of arbitrary callables
```

Comment:
Hi @Scienfitz & @RimRihana, I did quite a major review / change of the PR content, thus I'd ask to review from scratch, please. The things I've change (roughly):
* The example had waaaay to many layers of complexity to just convey the principle, IMO (e.g. no complex recommenders needed, not even a campaign/objective, no iteration needed, ...). Plus the text did not match what happened in the code blocks. Hence, I redesigned it from scratch such that it really contains the minimal logic only.
* One the one hand, there were some uncessary tests, on the other hand, there were missing tests that still resulted in failing cases. Hence, refactored the tests and adjusted the logic to catch the failing cases.
* I extended the registration mechanism: 1) now, you can provide "pre" and "post" hooks (see example why both is useful) 2) You can provide as many hooks per class as you wish
* The example now shows two different hooks of different complexity
* Renamed the example because the file name shows up in the docs

Please let me know if you have questions and give it a critical review 🙃

Reply:
(just noticed: forgot to update the docstring – will do so now and push the commit soon)

Reply:
Hi @AdrianSosic, thank's a lot for your effort! 
I did not think about having a hook list, because we could just register the hooks any time and one after the other (based on the use case), but this way looks more efficient. 
I just had one change suggestion that I will mention below :)

Reply:
There will be comments from me on this topic later, but I first want to talk about some stuff via meetings.